web-api/src/middleware/call_context.rs

```diff
@@ -0,0 +1,227 @@
+    /// accept additional context information.
+    ///
+    /// Be aware that in futures spawned from the main request handling future
+    /// this context will not be available. If you still need it there move a clone
+    /// of it to the spawned task and use [`ApiCallContext.wrap_future()`].
+    pub(crate) fn try_with_using_task_local<R>(
+        func: impl FnOnce(&Arc<ApiCallContext>) -> R,
+    ) -> Result<R, AccessError> {
+        API_CALL_CONTEXT.try_with(func).map_err(|_| AccessError {
+            method: "try_with_using_task_local",
+        })
+    }
+
+    /// Tries to return the current `ApiCallContext` based on a request.
+    ///
+    /// The context will be setup by the `setup_call_context` middleware.
+    pub(crate) fn try_with_using_request<R>(
+        request: &HttpRequest,
+        func: impl FnOnce(&Arc<ApiCallContext>) -> R,
+    ) -> Result<R, AccessError> {
+        let extensions = request.extensions();
+        Ok(func(extensions.get::<Arc<ApiCallContext>>().ok_or(
+            AccessError {
+                method: "try_with_using_request",
+            },
+        )?))
+    }
+
+    /// Wraps a future so that when it's polled the `ApiCallContext` will be available as task local data.
+    pub(crate) fn wrap_future<F>(
+        self: Arc<Self>,
+        future: F,
+    ) -> TaskLocalFuture<Arc<ApiCallContext>, F>
+    where
+        F: 'static + Future,
+    {
+        API_CALL_CONTEXT.scope(self, future)
+    }
+}
+
+#[derive(Debug, Error)]
+#[error("Can not access ApiCallContext from {method}")]
+pub(crate) struct AccessError {
+    method: &'static str,
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display, PartialEq, Deserialize, Serialize)]
+#[serde(transparent)]
+pub(crate) struct TenantId(Uuid);
+
+impl TenantId {
+    pub(crate) const fn missing() -> Self {
+        Self(Uuid::nil())
+    }
+}
+
+impl TryFrom<&'_ HeaderValue> for TenantId {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &HeaderValue) -> Result<Self, Self::Error> {
+        let header = value.to_str().map_err(|_| {
+            anyhow!(
+                "Non Utf-8 compatible {TENANT_ID_HEADER} header: {:?}",
+                String::from_utf8_lossy(value.as_bytes())
+            )
+        })?;
+        header
+            .parse::<Uuid>()
+            .map_err(|_| anyhow!("Non UUID {TENANT_ID_HEADER} header: {header}"))
+            .map(TenantId)
+    }
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display, Serialize)]
+#[serde(transparent)]
+pub(crate) struct RequestId(Uuid);
+
+impl RequestId {
+    pub(crate) fn generate() -> Self {
+        Self(Uuid::new_v4())
+    }
+
+    pub(crate) const fn missing() -> Self {
+        Self(Uuid::nil())
+    }
+}
+
+/// Configuration for handling tenants.
+#[derive(Debug, Default, Deserialize, Serialize)]
+#[serde(default)]
+pub struct Config {
+    /// Mocks the tenant id for legacy single tenant deployments
+    pub(crate) legacy_tenant: Option<TenantId>,
```

Comment:
I'm not convinced by this. This approach means that all the changes to the database and configuration in our environments must be done before this is merged.
We can just do nothing if the tenant id `None`

Reply:
> This approach means that all the changes to the database and configuration in our environments must be done before this is merged.

No it doesn't, it only means that a single config option has to be set per-tenant before merging this.

We can also set it initially to an arbitrary id and later set it to ids we care about, this won't be a problem.

The follow up PR creates a role for the legacy tenant on the fly on startup if it's missing, but in difference to normal multi tenant that role will still use the `public` schema.

This also means that for legacy tenants we still can change the ids on the fly as they don't appear anywhere in the data for now (through they do appear in the logs).

Reply:
done now config defaults to legacy tenant with an  `Uuid::nill()` as tenant id

other ids still can be set, legacy tenant can be disabled by setting `use_legacy_tenant = false` (`use_legacy_tenant = true` is the default and `use_legacy_tenant = "uuid...."` is for when we start giving non-multi-tenant tenants properly tenant ids and future migrations).